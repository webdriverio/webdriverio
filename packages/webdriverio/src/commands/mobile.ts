export * from './mobile/dragAndDrop.js'
export * from './mobile/longPress.js'
export * from './mobile/pinch.js'
export * from './mobile/zoom.js'
