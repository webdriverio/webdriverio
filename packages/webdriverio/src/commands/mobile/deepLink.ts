/**
 *
 * Open a deep link URL in the mobile app based on the url and the app's package name (Android) or bundle ID (iOS).
 *
 * <example>
    :deeplink.js
    it('should open a deep link for the WDIO native demo app', async () => {
        // open the Drag tab with a deep link (this the bundleId for the iOS Demo App)
        await browser.deepLink('wdio://drag', 'org.reactjs.native.example.wdiodemoapp');

        // Or open the Drag tab with a deep link (this the package name for the Android Demo App)
        await browser.deepLink('wdio://drag', 'com.wdiodemoapp');

        // Or if you want to have it "cross-platform" you can use it like this
        await browser.deepLink('wdio://drag', browser.isIOS ? 'org.reactjs.native.example.wdiodemoapp' : 'com.wdiodemoapp');
    })
 * </example>
 *
 * @param {string}  link            The deep link URL that should be opened in the mobile app. It should be a valid deep link URL (e.g. `myapp://path`). If it's a universal deep link, which can be used for iOS, use the `browser.url("your-url")`-method.
 * @param {string}  appIdentifier   The value of the `package` (Android) or `bundleId` (iOS) of the app that the deep link should open.
 */
export async function deepLink(
    this: WebdriverIO.Browser,
    link: string,
    appIdentifier: string
) {
    const browser = this

    if (!browser.isMobile) {
        throw new Error('The `deepLink` command is only available for mobile platforms.')
    }

    if (!isDeepLinkUrl(link)) {
        throw new Error(`The provided link is not a valid deep link URL.${browser.isIOS ? ' If your url is a `universal deep link` then use the `url` command instead.' : ''}`)
    }

    if (!appIdentifier) {
        const mobileOS = browser.isIOS ? 'iOS' : 'Android'
        const identifierValue = browser.isIOS ? 'bundleId' : 'package'

        throw new Error(`When using a deep link URL for ${mobileOS}, you need to provide the \`${identifierValue}\` of the app that the deep link should open.`)
    }

    return browser.execute('mobile:deepLink', {
        url: link,
        [browser.isIOS ? 'bundleId' : 'package']: appIdentifier,
    })
}

function isDeepLinkUrl(link: string): boolean {
    const deepLinkRegex = /^(?!https?:\/\/)[a-zA-Z][\w+\-.]*:\/\//

    return deepLinkRegex.test(link)
}
