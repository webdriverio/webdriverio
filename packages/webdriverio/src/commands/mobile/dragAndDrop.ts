/**
 *
 * Drag an item to a destination element or position.
 *
 * :::info
 *
 * The functionality of this command highly depends on the way drag and drop is
 * implemented in your app. If you experience issues please post your example
 * in [#4134](https://github.com/webdriverio/webdriverio/issues/4134).
 *
 * Also make sure that the element you are dragging and the target where you are dropping are both visible on the screen.
 *
 * :::
 *
 * <example>
    :example.test.js
    it('should demonstrate the dragAndDrop command', async () => {
        const elem = $('#someElem')
        const target = $('#someTarget')

        // drag and drop to other element
        await elem.dragAndDrop(target)

        // drag and drop relative from current position
        await elem.dragAndDrop({ x: 100, y: 200 })
    })
 * </example>
 *
 * @alias element.dragAndDrop
 * @param {Element|DragAndDropCoordinate}   target            destination element or object with x and y properties
 * @param {DragAndDropOptions=}             options           dragAndDrop command options
 * @param {Number=}                         options.duration  how long the drag should take place
 * @mobileElement
 */
// actual implementation is located in packages/webdriverio/src/element/dragAndDrop.ts
