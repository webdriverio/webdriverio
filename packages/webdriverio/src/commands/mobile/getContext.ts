import logger from '@wdio/logger'
import type { Context, DetailedContext } from '@wdio/protocols'

import type { AppiumDetailedCrossPlatformContexts } from '../../types.js'

const log = logger('webdriver')

/**
 * Retrieve the context of the current session.
 *
 * This method enhances the default Appium `context`/WebdriverIO `getContext` command by providing an option to
 * return detailed context information, making it easier to work with hybrid apps that use webviews.
 *
 * ### How Contexts Work
 * Refer to [Hybrid Apps documentation](/docs/api/mobile#hybrid-apps) for more information. Below is an explanation of the challenges associated with the `getContext` command:
 *
 * #### For Android:
 * - Webviews can contain multiple pages (like browser tabs), and identifying the correct page requires additional metadata
 *   such as `title` or `url`.
 * - The default Appium methods only provide basic context names (e.g., `WEBVIEW_{packageName}`) without detailed information
 *   about the pages inside the webview.
 *
 * #### For iOS:
 * - Each webview is identified by a generic `WEBVIEW_{id}` string, which doesn’t indicate its contents or the app screen
 *   it belongs to.
 *
 * ### Why Use This Method?
 * - **Default Behavior**:
 *   - Returns the current context as a string (e.g., `NATIVE_APP` or `WEBVIEW_{id}`).
 * - **Detailed Context**:
 *   - When `returnDetailedContext` is enabled, retrieves metadata such as:
 *     - **Android**: `packageName`, `title`, `url`, and `webviewPageId`.
 *     - **iOS**: `bundleId`, `title`, and `url`.
 * - **Android-Specific Options**:
 *   - Retry intervals and timeouts can be customized to handle delays in webview initialization.
 *
 * :::info Notes and Limitations
 *
 * - If `returnDetailedContext` is not enabled, the method behaves like the default Appium `getContext` method.
 * - If you want to use the "default" Appium `context` method, you can use the `driver.getAppiumContext()` method, see
 * also the [Appium Contexts](/docs/api/appium#getappiumcontext) command.
 * - **Android:** Android-specific options (`androidWebviewConnectionRetryTime` and `androidWebviewConnectTimeout`) have no effect on iOS.
 * - Logs warnings if multiple or no detailed contexts are found:
 *   - `We found more than 1 detailed context for the current context '{context}'. We will return the first context.`
 *   - `We did not get back any detailed context for the current context '{context}'. We will return the current context as a string.`
 *
 * :::
 *
 * <example>
    :default.test.js
    it('should return the current context with the default Appium `context` method', async () => {
        // For Android
        await driver.getContext()
        // Returns 'WEBVIEW_com.wdiodemoapp' or 'NATIVE_APP'
        //
        // For iOS, the context will be 'WEBVIEW_{number}'
        await driver.getContext()
        // Returns 'WEBVIEW_94703.19' or 'NATIVE_APP'
    })
 * </example>
 *
 * <example>
    :detailed.test.js
    it('should return the context of the current session with more detailed information', async () => {
        // For Android
        await driver.getContext({ returnDetailedContext: true})
        // Returns or `NATIVE_APP`, or
        // {
        //   id: 'WEBVIEW_com.wdiodemoapp',
        //   title: 'WebdriverIO · Next-gen browser and mobile automation test framework for Node.js | WebdriverIO',
        //   url: 'https://webdriver.io/',
        //   packageName: 'com.wdiodemoapp',
        //   webviewPageId: '5C0425CF67E9B169245F48FF21172912'
        // }
        //
        // For iOS, the context will be 'WEBVIEW_{number}'
        await driver.getContext({ returnDetailedContext: true})
        // Returns or `NATIVE_APP`, or
        // {
        //   id: 'WEBVIEW_64981.1',
        //   title: 'WebdriverIO · Next-gen browser and mobile automation test framework for Node.js | WebdriverIO',
        //   url: 'https://webdriver.io/',
        //   bundleId: 'org.reactjs.native.example.wdiodemoapp'
        // }
    })
 * </example>
 *
 * <example>
    :customize.retry.test.js
    it('should be able to cusomize the retry intervals and timeouts to handle delayed webview initialization', async () => {
        // For Android
        await driver.getContext({
            returnDetailedContext: true,
            // NOTE: The following options are Android-specific
            // For Android we might need to wait a bit longer to connect to the webview, so we can provide some additional options
            androidWebviewConnectionRetryTime: 1*1000,  // Retry every 1 second
            androidWebviewConnectTimeout: 10*1000,      // Timeout after 10 seconds
        })
    })
 * </example>
 *
 * @param {GetContextsOptions=} options                                     The `getContext` options (optional)
 * @param {boolean=}            options.returnDetailedContext               By default, we only return the context name based on the default Appium `context` API, which is only a string. If you want to get back detailed context information, set this to `true`. Default is `false` (optional).
 * @param {number=}             options.androidWebviewConnectionRetryTime   The time in milliseconds to wait between each retry to connect to the webview. Default is `500` ms (optional). <br /><strong>ANDROID-ONLY</strong>
 * @param {number=}             options.androidWebviewConnectTimeout        The maximum amount of time in milliseconds to wait for a web view page to be detected. Default is `5000` ms (optional). <br /><strong>ANDROID-ONLY</strong>
 * @skipUsage
 */
export async function getContext(
    this: WebdriverIO.Browser,
    options?: {
        returnDetailedContext?: boolean,
        androidWebviewConnectionRetryTime?: number,
        androidWebviewConnectTimeout?: number,
    }
): Promise<string | DetailedContext> {
    const browser = this

    if (!browser.isMobile) {
        throw new Error('The `getContext` command is only available for mobile platforms.')
    }

    const currentAppiumContext: Context = await browser.getAppiumContext()

    if ((!options || !options?.returnDetailedContext) || currentAppiumContext === 'NATIVE_APP') {
        return currentAppiumContext
    }

    delete options.returnDetailedContext

    return getDetailedContext(browser, currentAppiumContext as string, options)
}

async function getDetailedContext(
    browser: WebdriverIO.Browser,
    currentAppiumContext: string,
    options?: { androidWebviewConnectionRetryTime?: number, androidWebviewConnectTimeout?: number },
): Promise<string | DetailedContext> {
    const detailedContexts = await browser.getContexts({
        ...{ options },
        // Defaults
        returnDetailedContexts: true,           // We want to get back the detailed context information
        isAndroidWebviewVisible: true,          // We only want to get back the visible webviews
        filterByCurrentAndroidApp: true,        // We only want to get back the webviews that are attached to the current app
        returnAndroidDescriptionData: false,    // We don't want to get back the Android Webview description data
    }) as AppiumDetailedCrossPlatformContexts
    const parsedContexts = detailedContexts.filter((context) => context.id === currentAppiumContext) as DetailedContext[]

    if (parsedContexts.length > 1) {
        log.warn(`We found more than 1 detailed context for the current context '${currentAppiumContext}'. We will return the first context.`)

        return parsedContexts[0]
    } else if (parsedContexts.length === 0) {
        log.warn(`We did not get back any detailed context for the current context '${currentAppiumContext}'. We will return the current context as a string.`)

        return currentAppiumContext
    }

    return parsedContexts[0]
}
