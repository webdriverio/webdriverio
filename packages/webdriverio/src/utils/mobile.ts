import type { Capabilities } from '@wdio/types'
import type { PinchAndZoomOptions } from '../types.js'

const appiumKeys = ['app', 'bundleId', 'appPackage', 'appActivity', 'appWaitActivity', 'appWaitPackage'] as const
type AppiumKeysType = typeof appiumKeys[number]

export function getNativeContext({ capabilities, isMobile }:
    { capabilities: WebdriverIO.Capabilities, isMobile: boolean }
): boolean {
    if (!capabilities || typeof capabilities !== 'object' || !isMobile) {
        return false
    }

    const isAppiumAppCapPresent = (capabilities: Capabilities.RequestedStandaloneCapabilities) => {
        return appiumKeys.some((key) => (
            (capabilities as Capabilities.AppiumCapabilities)[key as keyof Capabilities.AppiumCapabilities] !== undefined ||
            (capabilities as WebdriverIO.Capabilities)['appium:options']?.[key as AppiumKeysType] !== undefined)
        )
    }
    const isBrowserNameFalse = !!capabilities?.browserName === false
    // @ts-expect-error
    const isAutoWebviewFalse = capabilities?.autoWebview !== true

    return isBrowserNameFalse && isAppiumAppCapPresent(capabilities) && isAutoWebviewFalse
}

export function getMobileContext({ capabilities, isAndroid, isNativeContext }:
    { capabilities: WebdriverIO.Capabilities, isAndroid: boolean, isNativeContext: boolean }
): string | undefined {
    return isNativeContext ? 'NATIVE_APP' :
    // Android webviews are always WEBVIEW_<package_name>, Chrome will always be CHROMIUM
    // We can only determine it for Android and Chrome, for all other, including iOS, we return undefined
        isAndroid && capabilities?.browserName?.toLowerCase() === 'chrome' ? 'CHROMIUM' :
            undefined
}

export function calculateAndroidPinchAndZoomSpeed({ browser, duration, scale }:
    { browser: WebdriverIO.Browser, duration: number, scale: number }
): number {
    // @ts-expect-error
    const deviceScreenSize = (browser.capabilities?.deviceScreenSize || '1080x2400').split('x').reduce((a, b) => a * b)
    // Calculate base distance (diagonal of the screen in physical pixels)
    const baseDistance = Math.sqrt(deviceScreenSize)
    // Adjust gesture distance by scale (ensure meaningful gestureDistance even for low scales)
    const gestureDistance = Math.max(baseDistance * Math.abs(scale), baseDistance * 0.1) // Minimum 10% distance
    const durationSeconds = duration / 1000

    return Math.floor(gestureDistance / durationSeconds)
}

export function validatePinchAndZoomOptions({ browser, gesture, options }:
    { browser: WebdriverIO.Browser, gesture: 'pinch' | 'zoom', options: Partial<PinchAndZoomOptions> }
): { scale: number, duration: number } {
    if (typeof options !== 'undefined' && (typeof options !== 'object' || Array.isArray(options))) {
        throw new TypeError('Options must be an object')
    }

    const DEFAULT_SCALE = 0.5
    const DEFAULT_DURATION = browser.isIOS ? 1.5 : 1500
    const MIN_SCALE = 0
    const MAX_SCALE = 1
    const MIN_DURATION_MS = 500
    const MAX_DURATION_MS = 10000
    const { scale: scaleOption, duration: durationOption } = options

    const scale = typeof scaleOption === 'number'
        ? scaleOption >= MIN_SCALE && scaleOption <= MAX_SCALE
            ? scaleOption
            : (() => { throw new Error(`The 'scale' option must be a number between ${MIN_SCALE} and ${MAX_SCALE}`) })()
        : DEFAULT_SCALE
    const duration = typeof durationOption === 'number'
        ? durationOption >= MIN_DURATION_MS && durationOption <= MAX_DURATION_MS
            ? browser.isIOS ? durationOption / 1000 : durationOption
            : (() => { throw new Error(`The 'duration' option must be between ${MIN_DURATION_MS} and ${MAX_DURATION_MS} ms (${MIN_DURATION_MS/1000} and ${MAX_DURATION_MS/1000}  seconds)`) })()
        : DEFAULT_DURATION

    return {
        duration,
        scale: browser.isIOS && gesture === 'zoom' ? scale * 10 : scale,
    }
}
