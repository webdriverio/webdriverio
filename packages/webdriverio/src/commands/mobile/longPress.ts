import { getBrowserObject } from '@wdio/utils'
import type { LongPressOptions } from '../../types.js'

/**
 *
 * Performs a long press gesture on the given element on the screen.
 *
 * This issues a WebDriver `action` command for the selected element. It is based on the `click` command.
 *
 * <example>
    :longpress.offset.js
    it('should demonstrate a longPress using an offset on the iOS Contacts icon', async () => {
        const contacts = $('~Contacts')
        // opens the Contacts menu on iOS where you can quickly create
        // a new contact, edit your home screen, or remove the app
        // clicks 30 horizontal and 10 vertical pixels away from location of the icon (from center point of element)
        await contacts.longPress({ x: 30, y: 10 })
    })
 * </example>
 *
 * <example>
    :longpress.example.js
    it('should be able to open the contacts menu on iOS by executing a longPress of 5 seconds', async () => {
        const contacts = $('~Contacts')
        // opens the Contacts menu on iOS where you can quickly create
        // a new contact, edit your home screen, or remove the app
        await contacts.longPress({ duration: 5 * 1000 })
    })
 * </example>
 *
 * @alias element.click
 * @uses protocol/element, protocol/elementIdClick, protocol/performActions, protocol/positionClick
 * @type action
 * @param {LongPressOptions=} options           Long press options (optional)
 * @param {number=}           options.x         Number (optional)
 * @param {number=}           options.y         Number (optional)
 * @param {number=}           options.duration  Duration of the press in ms, default is 1500 ms <br /><strong>MOBILE-ONLY</strong>
 * @mobileElement
 */
export function longPress(
    this: WebdriverIO.Element,
    options?: Partial<LongPressOptions>
) {
    const browser = getBrowserObject(this)

    if (!browser.isMobile) {
        throw new Error('The longPress command is only available for mobile platforms.')
    }

    if (
        typeof options !== 'undefined' &&
        (typeof options !== 'object' || Array.isArray(options))
    ) {
        throw new TypeError('Options must be an object')
    }

    const defaultOptions: LongPressOptions = {
        duration: 1500,
        x: 0,
        y: 0,
    }

    const { duration, x, y }: LongPressOptions = { ...defaultOptions, ...options }

    return this.click({ duration, x, y })
}
