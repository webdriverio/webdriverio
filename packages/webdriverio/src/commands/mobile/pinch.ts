import { getBrowserObject } from '@wdio/utils'
import type { PinchAndZoomOptions } from '../../types.js'
import { calculateAndroidPinchAndZoomSpeed, validatePinchAndZoomOptions } from '../../utils/mobile.js'

/**
 *
 * Performs a pinch gesture on the given element on the screen.
 *
 * :::info
 *
 * Pinching is done based on native mobile gestures. It is only supported for the following drivers:
 * - [appium-uiautomator2-driver](https://github.com/appium/appium-uiautomator2-driver/blob/master/docs/android-mobile-gestures.md#mobile-pinchclosegesture) for Android
 * - [appium-xcuitest-driver](https://appium.github.io/appium-xcuitest-driver/latest/reference/execute-methods/#mobile-pinch) for iOS
 *
 * :::
 *
 * <example>
    :pinch.js
    it('should demonstrate a pinch on Google maps', async () => {
        const mapsElement = $('//*[@resource-id="com.google.android.apps.maps:id/map_frame"]')
        // Pinch with the default duration scale
        await mapsElement.pinch()
        // Pinch with a custom duration and scale
        await mapsElement.pinch({ duration: 4000, scale: 0.9 })
    })
 * </example>
 *
 * @alias element.pinch
 * @param {PinchOptions=} options           pinch options (optional)
 * @param {number=}       options.duration  The duration in millisecond of how fast the pinch should be executed, minimal is 500 ms and max is 10000 ms. The default is 1500 ms (1.5 seconds) (optional)
 * @param {number=}       options.scale     The scale of how big the pinch should be according to the screen. Valid values must be float numbers in range 0..1, where 1.0 is 100% (optional)
 * @mobileElement
 */
export async function pinch(
    this: WebdriverIO.Element,
    options: Partial<PinchAndZoomOptions> = {}
) {
    const browser = getBrowserObject(this)

    if (!browser.isMobile) {
        throw new Error('The pinch command is only available for mobile platforms.')
    }

    const { duration, scale } = validatePinchAndZoomOptions({ browser, gesture: 'pinch', options })

    const gestureConfig = browser.isIOS
        ? {
            elementId: await this.elementId,
            scale,
            velocity: -Math.abs(duration), // Velocity is always negative for iOS pinch
        }
        : {
            elementId: await this.elementId,
            percent: scale,
            speed: calculateAndroidPinchAndZoomSpeed({ browser, duration, scale }),
        }

    return browser.execute(browser.isIOS ? 'mobile: pinch' : 'mobile: pinchCloseGesture', gestureConfig)
}
