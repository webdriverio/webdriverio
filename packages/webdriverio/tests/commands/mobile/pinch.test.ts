import path from 'node:path'
import { expect, describe, it, vi, beforeEach } from 'vitest'
import { remote } from '../../../src/index.js'
import { calculateAndroidPinchAndZoomSpeed, validatePinchAndZoomOptions } from '../../../src/utils/mobile.js'

vi.mock('fetch')
vi.mock('@wdio/logger', () => import(path.join(process.cwd(), '__mocks__', '@wdio/logger')))
vi.mock(import('../../../src/utils/mobile.js'), async (importOriginal) => {
    const actual = await importOriginal()
    return {
        ...actual,
        calculateAndroidPinchAndZoomSpeed: vi.fn(),
        validatePinchAndZoomOptions: vi.fn(),
    }
})

describe('pinch test', () => {
    let browser: WebdriverIO.Browser
    let elem: WebdriverIO.Element

    beforeEach(async () => {
        vi.mocked(fetch).mockClear()
    })

    it('should throw an error for non-mobile platforms', async () => {
        browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar',
            } as any
        })
        // @ts-expect-error
        elem = await browser.$('#foo')
        expect(elem.pinch()).rejects.toThrow('The pinch command is only available for mobile platforms.')
    })

    it('should execute a pinch for iOS with the correct values', async () => {
        browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar',
                mobileMode: true,
                platformName: 'iOS',
            } as any
        })
        // @ts-expect-error
        elem = await browser.$('#foo')
        // @ts-expect-error mock feature
        elem.elementId = { pinch: 'mockFunction'  }
        vi.mocked(validatePinchAndZoomOptions).mockReturnValue({
            duration: 1500,
            scale: 0.8
        })

        await elem.pinch()

        const { calls } = vi.mocked(fetch).mock
        expect(calls).toHaveLength(3)
        const [[sessionCallUrl], [elementCallUrl], [executeCallUrl, executeCallOptions]] = calls as any
        expect(sessionCallUrl.pathname).toEqual('/session')
        expect(elementCallUrl.pathname).toEqual('/session/foobar-123/element')
        expect(executeCallUrl.pathname).toEqual('/session/foobar-123/execute/sync')
        expect(JSON.parse(executeCallOptions.body)).toMatchSnapshot()
    })

    it('should execute a pinch for Android with the correct values', async () => {
        browser = await remote({
            baseUrl: 'http://foobar.com',
            capabilities: {
                browserName: 'foobar',
                mobileMode: true,
                platformName: 'Android',
            } as any
        })
        // @ts-expect-error
        elem = await browser.$('#foo')
        // @ts-expect-error mock feature
        elem.elementId = { pinch: 'mockFunction'  }
        vi.mocked(validatePinchAndZoomOptions).mockReturnValue({
            duration: 1500,
            scale: 0.8
        })
        vi.mocked(calculateAndroidPinchAndZoomSpeed).mockReturnValue(888)

        await elem.pinch()

        const { calls } = vi.mocked(fetch).mock
        expect(calls).toHaveLength(3)
        const [[sessionCallUrl], [elementCallUrl], [executeCallUrl, executeCallOptions]] = calls as any
        expect(sessionCallUrl.pathname).toEqual('/session')
        expect(elementCallUrl.pathname).toEqual('/session/foobar-123/element')
        expect(executeCallUrl.pathname).toEqual('/session/foobar-123/execute/sync')
        expect(JSON.parse(executeCallOptions.body)).toMatchSnapshot()
    })
})
