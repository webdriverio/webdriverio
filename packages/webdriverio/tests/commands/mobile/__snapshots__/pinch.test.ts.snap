// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`pinch test > should execute a pinch for Android with the correct values 1`] = `
{
  "args": [
    {
      "elementId": {
        "pinch": "mockFunction",
      },
      "percent": 0.8,
      "speed": 888,
    },
  ],
  "script": "mobile: pinchCloseGesture",
}
`;

exports[`pinch test > should execute a pinch for iOS with the correct values 1`] = `
{
  "args": [
    {
      "elementId": {
        "pinch": "mockFunction",
      },
      "scale": 0.8,
      "velocity": -1500,
    },
  ],
  "script": "mobile: pinch",
}
`;
